pallets/rewards/src/mechanism.rs

```diff
@@ -0,0 +1,171 @@
+use frame_support::traits::tokens::Balance;
+use sp_runtime::{traits::Get, ArithmeticError};
+
+pub mod base;
+pub mod base_with_currency_movement;
+
+pub trait RewardMechanism {
```

Comment:
I've added some comments in my last commit, but not sure if they are a little bit redundant. I can not explain more because what happens depends on the implementation.